apm-agent-plugins/apm-log-shader-plugin/src/main/java/co/elastic/apm/agent/log/shader/log4j2/helper/Log4j2LogShadingHelper.java

```diff
@@ -0,0 +1,122 @@
+import co.elastic.apm.agent.log.shader.Utils;
+import co.elastic.logging.log4j2.EcsLayout;
+import org.apache.logging.log4j.core.appender.AbstractOutputStreamAppender;
+import org.apache.logging.log4j.core.appender.RollingRandomAccessFileAppender;
+import org.apache.logging.log4j.core.appender.rolling.DefaultRolloverStrategy;
+import org.apache.logging.log4j.core.appender.rolling.RolloverStrategy;
+import org.apache.logging.log4j.core.appender.rolling.SizeBasedTriggeringPolicy;
+import org.apache.logging.log4j.core.appender.rolling.TriggeringPolicy;
+import org.apache.logging.log4j.core.config.DefaultConfiguration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+import java.lang.reflect.Method;
+
+
+public class Log4j2LogShadingHelper extends AbstractLogShadingHelper<AbstractOutputStreamAppender<?>> {
+
+    public static final String ECS_SHADE_APPENDER = "EcsShadeAppender";
+
+    private static final Logger logger = LoggerFactory.getLogger(Log4j2LogShadingHelper.class);
+
+    public Log4j2LogShadingHelper(ElasticApmTracer tracer) {
+        super(tracer);
+    }
+
+    @Override
+    protected boolean isShadingAppender(AbstractOutputStreamAppender<?> appender) {
+        //noinspection StringEquality
+        return appender.getName() == ECS_SHADE_APPENDER;
+    }
+
+    @Override
+    @Nullable
+    protected AbstractOutputStreamAppender<?> createAndConfigureAppender(AbstractOutputStreamAppender<?> originalAppender) {
+
+        String logFile = null;
+
+        // Using class names and reflection in order to avoid version sensitivity
+        String appenderClassName = originalAppender.getClass().getName();
+        if (appenderClassName.equals("org.apache.logging.log4j.core.appender.FileAppender") ||
+            appenderClassName.equals("org.apache.logging.log4j.core.appender.RollingFileAppender") ||
+            appenderClassName.equals("org.apache.logging.log4j.core.appender.RandomAccessFileAppender") ||
+            appenderClassName.equals("org.apache.logging.log4j.core.appender.RollingRandomAccessFileAppender") ||
+            appenderClassName.equals("org.apache.logging.log4j.core.appender.MemoryMappedFileAppender")) {
+            try {
+                Method getFileNameMethod = originalAppender.getClass().getDeclaredMethod("getFileName");
+                logFile = (String) getFileNameMethod.invoke(originalAppender);
+            } catch (Exception e) {
+                logger.error("Failed to obtain log file name from file appender", e);
+            }
+        }
+
+        if (logFile == null) {
+            return null;
+        }
+
+        String shadeFile = Utils.computeShadeLogFilePath(logFile);
+
+        EcsLayout ecsLayout = EcsLayout.newBuilder()
+            .setServiceName(getServiceName())
+            .setIncludeMarkers(false)
+            .setIncludeOrigin(false)
+            .setStackTraceAsArray(true)
```

Comment:
Although this makes it more readable it probably has to be set to false so it works with the logs intake API which is currently spec'd to assume ND-JSON